### PR TITLE
Feat (ui): Add VAE Model to Recall Parameters

### LINKED ISSUE #4796: [bug]: VAE is not changing when "Use all"
### Is there an existing issue for this?

- [X] I have searched the existing issues

### OS

Windows

### GPU

cuda

### VRAM

11

### What version did you experience this issue on?

3.2.0rc3

### What happened?

When we press "Use All" on generated image - VAE what was used for generating of image is not setting.

But in different situations need either use old VAE from generated image or new VAE. It also will be good if there will flag like [FIX-VAE] - so that we can optionally enable or disable the VAE changing.
If it is hard for add - then let it changing as in generated image, please)

### Screenshots

_No response_

### Additional context

_No response_

### Contact Details

_No response_

### PR DESCRIPTION
**Pull request by StefanTobler**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [X] Feature
- [X] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission

## Have you discussed this change with the InvokeAI team?
- [ ] Yes
- [X] No, because: it is 2 am est

#4796      

## Have you updated all relevant documentation?
- [ ] Yes
- [X] No

I don't believe there is any documentation explicitly referencing the recall parameters.

## Description

This code changes allows users to recall the VAE model as a Recall Parameter. The VAE model is now part of the "recall all parameters" setting. There is room to add the VAE Precision to the metadata and also making that a recall parameter..

## Related Tickets & Documents

- Related Issue #4796 
- Closes #4796 

## QA Instructions, Screenshots, Recordings

1. Generate an image using a VAE other than default.
2. Select that generated image, change the VAE model, and select "Use All"
3. VAE model should change in the UI.
4. VAE model is available in "recall parameters" of the metadata section.

https://github.com/invoke-ai/InvokeAI/assets/22108632/04a52722-6faf-4ab4-b414-81806fe4aa54

## Added/updated tests?

- [ ] Yes
- [X] No : no frontend test suite

### CODE REVIEW

**Comment by StefanTobler** [comment]
This PR may not be considered "complete" as it does not recall the "default" VAE model if that is selected. I think that this would need the addition of adding the default VAE to the metadata, as of right now, if the VAE model is default it is not added in the generated images metadata.

**Comment by StefanTobler** [comment]
Last note for tonight, I also noticed that the naming for VAE is a bit inconsistent around the frontend, it seems it is either VAE or VAE Model, but they both refer to the same thing.

File: invokeai/frontend/web/src/features/parameters/hooks/useRecallParameters.ts [hunk 1]
```diff
@@ -65,6 +65,7 @@ import {
   setSeed,
   setSteps,
   setWidth,
+  vaeSelected,
```
Comment (StefanTobler) [inline invokeai/frontend/web/src/features/parameters/hooks/useRecallParameters.ts:68]: This should function should be renamed to `setVae` or `setVaeModel` for consistency sake. I can do that if the reviewers don't mind, but to keep the scope of this PR small, I am not jumping the gun.

Comment (psychedelicious) [reply invokeai/frontend/web/src/features/parameters/hooks/useRecallParameters.ts:68]: It's OK to leave it, actually the other functions should be named to [model events](https://redux.js.org/style-guide/#model-actions-as-events-not-setters), but we only started doing that recently and it's not a hard rule

**Comment by psychedelicious** [comment]
Yep as you noted, lets add some additional logic to check if the vae is nullish when recalling all, and if so, set it to null. 

Looking good otherwise!

Stretch goal: check the RTK Query cache when recalling the VAE to ensure it is actually available on the system. Happy to have this be a future change, and we should do the same for when we recall a main model.

**Comment by StefanTobler** [comment]
@psychedelicious added the check for nullish variables. It is a bit verbose imo, but I am not sure if it is worth making it harder to read to reduce the code. 

Also debating on adding similar logic to `ImageMetadataActions.tsx` so that `vae` is displayed as "Default" when nullish. I am ignorant as to when this would be used in the event that VAE is not present for a reason. I guess it seems strange to me to be making all the exceptions for this one field but not all the others when null. Maybe overthinking.

**Comment by StefanTobler** [comment]
I guess it lead me to question, if the VAE is undefined, it is okay to assume we oughta force the default VAE model on the parameters?

**Comment by StefanTobler** [comment]
image

**Comment by Millu** [comment]
> I guess it lead me to question, if the VAE is undefined, it is okay to assume we oughta force the default VAE model on the parameters?

This behavior makes sense to me

**Comment by StefanTobler** [comment]
Agreed, I added that in the latest commit.

File: invokeai/frontend/web/src/features/parameters/hooks/useRecallParameters.ts [hunk 2]
```diff
@@ -87,8 +88,10 @@ import {
   isValidSteps,
   isValidStrength,
   isValidWidth,
+  isValidVaeModel,
   isValidBoolean,
 } from '../types/parameterSchemas';
+import _ from 'lodash';
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/parameters/hooks/useRecallParameters.ts:94]: We need to import from `lodash-es` to take advantage of tree-shaking and reduce bundle size

```suggestion
import { isNil } from 'lodash-es';
```

File: invokeai/frontend/web/src/features/parameters/hooks/useRecallParameters.ts [hunk 3]
```diff
@@ -798,6 +821,13 @@ export const useRecallParameters = () => {
       if (isValidScheduler(scheduler)) {
         dispatch(setScheduler(scheduler));
       }
+      if (isValidVaeModel(vae) || _.isNil(vae)) {
+        if (_.isNil(vae)) {
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/parameters/hooks/useRecallParameters.ts:825]:
```suggestion
        if (isNil(vae)) {
```

Comment (psychedelicious) [reply invokeai/frontend/web/src/features/parameters/hooks/useRecallParameters.ts:825]: there are a couple other occurrences

**Review by psychedelicious — APPROVED** [review]
Thanks @StefanTobler !

**Pull request merged**